### PR TITLE
[oneDNN] Quantized batch-matmul and fusions

### PR DESCRIPTION
**Pull request by mdfaijul**
This PR enables support for quantized batched matrices in the BatchMatMul op. A few binary ops can also be fused to it. Transformers based models frequently uses it and the fusions (e.g., BatchMatMul + Mul + Add) in their self-attention.

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @mdfaijul Can you please rebase your branch and resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

File: tensorflow/core/kernels/dequantize_op.cc [hunk 1]
```diff
@@ -70,12 +70,6 @@ class DequantizeOp : public OpKernel {
                   errors::InvalidArgument("Mode string must be 'MIN_COMBINED',"
                                           " 'MIN_FIRST', or 'SCALED', is '" +
                                           mode_string + "'"));
-    } else {
```
Comment (cantonios) [inline tensorflow/core/kernels/dequantize_op.cc:73]: Are you sure this isn't breaking something on CPU (e.g. without onednn)?

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc [hunk 2]
```diff
@@ -35,8 +35,10 @@ limitations under the License.
 #include "tensorflow/core/kernels/matmul_op_impl.h"
 #include "tensorflow/core/kernels/mkl/mkl_batch_matmul_helper.h"
 #include "tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h"
+#include "tensorflow/core/platform/errors.h"
 #include "tensorflow/core/platform/types.h"
 #include "tensorflow/core/util/matmul_bcast.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc:41]: I don't think this header exists anymore - it should just be "unsupported/Eigen..."

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 3]
```diff
@@ -809,11 +809,14 @@ class MklMatMulPrimitive : public MklPrimitive {
     return context_.prim_desc->scratchpad_desc();
   }
   void Execute(const std::shared_ptr<stream>& stream, const Tlhs* a_data,
-               const Trhs* b_data, const Toutput* c_data, void* sp_data,
-               void* mul_data = nullptr, void* add_data = nullptr) {
+               const Trhs* b_data, const Toutput* c_data,
+               const MklMatMulParams& matmul_params, void* sp_data,
+               const std::vector<void*> binary_op_fusions_data = {}) {
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:814]: Are you sure you want to pass a vector by value?

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 4]
```diff
@@ -823,10 +826,9 @@ class MklMatMulPrimitive : public MklPrimitive {
         static_cast<void*>(const_cast<Toutput*>(c_data)), *stream);
     context_.sp_mem->set_data_handle(sp_data, *stream);
 
-    if (mul_data != nullptr)
-      context_.mul_mem->set_data_handle(mul_data, *stream);
-    if (add_data != nullptr)
-      context_.add_mem->set_data_handle(add_data, *stream);
+    for (int i = 0; i < num_post_ops_data; ++i)
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:829]: Please use braces

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 5]
```diff
@@ -835,18 +837,29 @@ class MklMatMulPrimitive : public MklPrimitive {
...
-
-    // After execution, set data handle back
-    context_.a_mem->set_data_handle(DummyData);
     context_.b_mem->set_data_handle(DummyData);
     context_.c_mem->set_data_handle(DummyData);
     context_.sp_mem->set_data_handle(DummyData);
-    if (mul_data != nullptr) context_.mul_mem->set_data_handle(DummyData);
-    if (add_data != nullptr) context_.add_mem->set_data_handle(DummyData);
+    for (int i = 0; i < num_post_ops_data; ++i)
+      context_.post_ops_mem[i]->set_data_handle(DummyData);
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:862]: braces

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 6]
```diff
@@ -860,10 +873,15 @@ class MklMatMulPrimitive : public MklPrimitive {
     std::shared_ptr<dnnl::memory> a_mem;
     std::shared_ptr<dnnl::memory> b_mem;
     std::shared_ptr<dnnl::memory> c_mem;
-    std::shared_ptr<dnnl::memory> mul_mem;
-    std::shared_ptr<dnnl::memory> add_mem;
     std::shared_ptr<dnnl::memory> sp_mem;
 
+    // Quantization scale related memory
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:878]: nit: punctuation

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 7]
```diff
@@ -874,8 +892,11 @@ class MklMatMulPrimitive : public MklPrimitive {
     std::shared_ptr<dnnl::memory::desc> a_md;
     std::shared_ptr<dnnl::memory::desc> b_md;
     std::shared_ptr<dnnl::memory::desc> c_md;
-    std::shared_ptr<dnnl::memory::desc> mul_md;
-    std::shared_ptr<dnnl::memory::desc> add_md;
+
+    // Quantization scale related memory descriptors
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:896]: nit: punctuation

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 8]
```diff
@@ -926,29 +965,71 @@ class MklMatMulPrimitive : public MklPrimitive {
     auto const& post_op_params = params.post_op_params;
     dnnl::primitive_attr post_ops_attr;
     dnnl::post_ops post_ops;
+    std::unordered_map<string, bool> is_scale_set;
+    int binary_post_ops_count = 0;  // Keep track op binary fusions
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:969]: nit: punctuation on comment

**Review by cantonios — CHANGES_REQUESTED** [review]

**Comment by mdfaijul** [comment]
@cantonios @penpornk This PR has been out of sync for long time. I am working on this PR again. I have temporarily changed the status to draft and will make it ready for review shortly.

File: tensorflow/core/kernels/dequantize_op.cc (hunk 1 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/dequantize_op.cc]: Thanks! Reverted the change.

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc (hunk 2 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc]: Corrected.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h (hunk 4 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Corrected.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h (hunk 5 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:870]: Corrected.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h (hunk 6 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Corrected.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h (hunk 7 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Corrected.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h (hunk 8 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Corrected.

**Comment by mdfaijul** [comment]
@cantonios @penpornk I have addressed review comments. Thank you very much for the suggestions.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h (hunk 3 above)
Comment (cantonios) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: This is still unresolved.  You shouldn't really be creating empty vectors for default arguments here, and I don't think you want to be passing vectors by value.  I suggest either refactoring so that all calls pass in a vector by const-reference, or if you want to keep an "optional" parameter, writing two overloads.

Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: I have changed it to const reference and and removed default arguments. Thanks!

**Review by cantonios — APPROVED** [review]

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 9]
```diff
@@ -1922,6 +1922,39 @@ REGISTER_OP("_MklLayerNorm")
...
+    .Attr("Tdevice_outputs: list(type) >= 0 = []")
+    .Attr("Thost_outputs: list(type) >= 0 = []")
+    // The following attributes T1, T2, Tfusion, and Tout are members of Tinputs
+    // and Toutputs, used here for type constraints in the templatized OpKernel
+    // registrations.
+    .Attr("T1: quantizedtype")  // 0-th input    
+    .Attr("T2: quantizedtype")  // 1st input
+    // Currently, restricting all additional inputs of same type.
+    .Attr("U: {bfloat16, float, quantizedtype} = DT_FLOAT")
+    .Attr("Tout: {bfloat16, float, quantizedtype} = DT_FLOAT")  // 0-th output
```
Comment (cantonios) [inline tensorflow/core/ops/mkl_nn_ops.cc:1948]: There are some trailing spaces in this file.  Can you run a formatter or remove them all?

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h [hunk 10]
```diff
@@ -19,6 +19,7 @@ limitations under the License.
 #ifdef INTEL_MKL
 
 #include "dnnl.hpp"
+#include "tensorflow/core/framework/types.pb.h"
```
Comment (cantonios) [inline tensorflow/core/kernels/mkl/mkl_kernel_util.h:22]: Apparently we can't import protobuf headers in kernels.  Why is this needed?  Is there a potential work-around?

**Review by cantonios — CHANGES_REQUESTED** [review]

**Review by cantonios — APPROVED** [review]

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h (hunk 10 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_kernel_util.h]: Corrected.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 9 above)
Comment (mdfaijul) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Corrected.

**Comment by penpornk** [comment]
This PR also caused Windows test failures on Broadwell.

```
==================== Test output for //tensorflow/core/kernels/mkl:onednn_fused_batchmatmul_op_test:
I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
[==========] Running 7 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 7 tests from Test/FusedBatchMatMulOpTest/0, where TypeParam = float
[ RUN      ] [Test/FusedBatchMatMulOpTest/0](http://test/FusedBatchMatMulOpTest/0).MulFusion
2024-01-30 20:16:30.660017: I [tensorflow/core/platform/cpu_feature_guard.cc:210](https://cs.corp.google.com/piper///depot/google3/tensorflow/core/platform/cpu_feature_guard.cc?l=210&cl=602747753)] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE SSE2 SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
[       OK ] Test/FusedBatchMatMulOpTest/0.MulFusion (129 ms)
[ RUN      ] Test/FusedBatchMatMulOpTest/0.AddFusion
[       OK ] Test/FusedBatchMatMulOpTest/0.AddFusion (65 ms)
[ RUN      ] Test/FusedBatchMatMulOpTest/0.MulAddFusion
[       OK ] Test/FusedBatchMatMulOpTest/0.MulAddFusion (55 ms)
[ RUN      ] Test/FusedBatchMatMulOpTest/0.QuantizedNoFusion
tensorflow/core/framework/tensor_testutil.cc(184): error: Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (0.051306240260601044 not close to 0.01457531750202179)
Expected: true
i = 0 Tx[i] = 0.051306240260601044 Ty[i] = 0.01457531750202179
tensorflow/core/framework/tensor_testutil.cc(184): error: Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (0.29469430446624756 not close to 0.40220573544502258)
Expected: true
i = 2 Tx[i] = 0.29469430446624756 Ty[i] = 0.40220573544502258
tensorflow/core/framework/tensor_testutil.cc(184): error: Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (0.075151577591896057 not close to 0.1325015127658844)
Expected: true
i = 4 Tx[i] = 0.075151577591896057 Ty[i] = 0.1325015127658844
tensorflow/core/framework/tensor_testutil.cc(184): error: Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (-0.32194021344184875 not close to -0.6212310791015625)
Expected: true
i = 5 Tx[i] = -0.32194021344184875 Ty[i] = -0.6212310791015625
tensorflow/core/framework/tensor_testutil.cc(184): error: Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (-0.13619852066040039 not close to -0.34168317914009094)
Expected: true
i = 6 Tx[i] = -0.13619852066040039 Ty[i] = -0.34168317914009094
tensorflow/core/framework/tensor_testutil.cc(184): error: Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (-0.36221140623092651 not close to -0.79298555850982666)
Expected: true
i = 7 Tx[i] = -0.36221140623092651 Ty[i] = -0.79298555850982666
tensorflow/core/framework/tensor_testutil.cc(184): error: Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (0.23144723474979401 not close to 0.088897265493869781)
Expected: true
i = 8 Tx[i] = 0.23144723474979401 Ty[i] = 0.088897265493869781
tensorflow/core/framework/tensor_testutil.cc(184): error: Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (0.096353642642498016 not close to 0.37416574358940125)
Expected: true
.
.
.
[==========] 7 tests from 1 test suite ran. (2437 ms total)
[  PASSED  ] 3 tests.
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] Test/FusedBatchMatMulOpTest/0.QuantizedNoFusion, where TypeParam = float
[  FAILED  ] Test/FusedBatchMatMulOpTest/0.QuantizedMulFusion, where TypeParam = float
[  FAILED  ] Test/FusedBatchMatMulOpTest/0.QuantizedAddFusion, where TypeParam = float
[  FAILED  ] Test/FusedBatchMatMulOpTest/0.QuantizedMulAddFusion, where TypeParam = float

 4 FAILED TESTS
```

**Comment by mdfaijul** [comment]
@penpornk @cantonios I have disabled the quantized tests on Windows. Let me know if it is okay.

**Review by penpornk — CHANGES_REQUESTED** [review]
I don't think we should disable the test. I was considering disabling the test in https://github.com/tensorflow/tensorflow/pull/56776 because it's an existing failure (failing without the PR) and it was the first failure I saw. Now that this PR and https://github.com/tensorflow/tensorflow/pull/56792 also have many failing tests, it's definitely concerning. 

I think we should disable the failing op fusions (graph rewrites) on Windows.

**Comment by mdfaijul** [comment]
@penpornk I have restricted the test to run on CPUs that has support for VNNI/AMX INT8. Please let me know if the PR still has more issues. Thanks!

**Review by cantonios — APPROVED** [review]

**Comment by elfringham** [comment]
This is causing many unit test failures on AARCH64. See https://github.com/tensorflow/tensorflow/actions/runs/7791926554/job/21248977277

**Comment by mdfaijul** [comment]
> This is causing many unit test failures on AARCH64. See https://github.com/tensorflow/tensorflow/actions/runs/7791926554/job/21248977277

@elfringham Thanks for reporting the issue. I am not familiar with AARCH64 code. Could anybody (familiar with AARCH64) help fixing the issue.

**Comment by elfringham** [comment]
```
[       OK ] LinalgGradTest.Einsum_Transpose (188 ms)
[ RUN      ] LinalgGradTest.Einsum_TransposeBroadcast
[       OK ] LinalgGradTest.Einsum_TransposeBroadcast (16 ms)
[ RUN      ] LinalgGradTest.Einsum_MatMul
--Type <RET> for more, q to quit, c to continue without paging--

Thread 447 "gradients_linal" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xfff7bf7fca00 (LWP 723392)]
dnnl::primitive_desc_base::query_md (this=0x0, what=dnnl::query::weights_md, idx=0)
    at external/mkl_dnn_acl_compatible/include/oneapi/dnnl/dnnl.hpp:4208
4208	external/mkl_dnn_acl_compatible/include/oneapi/dnnl/dnnl.hpp: No such file or directory.
(gdb) dir /home/andrew/src/tf_test/tensorflow-git/bazel-ci_build-cache/.cache/bazel/_bazel_andrew/eab0d61a99b6696edb3d2aff87b585e8
Source directories searched: /home/andrew/src/tf_test/tensorflow-git/bazel-ci_build-cache/.cache/bazel/_bazel_andrew/eab0d61a99b6696edb3d2aff87b585e8:$cdir:$cwd
(gdb) l
4203	                    [=](query q) { return what == q; }))
4204	            DNNL_THROW_ERROR(dnnl_invalid_arguments,
4205	                    "memory descriptor query is invalid");
4206	
4207	        const_dnnl_memory_desc_t cdesc = dnnl_primitive_desc_query_md(
4208	                get(), dnnl::convert_to_c(what), idx);
4209	        if (!cdesc) return memory::desc();
4210	
4211	        dnnl_memory_desc_t cloned_md = nullptr;
4212	        error::wrap_c_api(dnnl_memory_desc_clone(&cloned_md, cdesc),
(gdb) bt
#0  dnnl::primitive_desc_base::query_md (this=0x0, what=dnnl::query::weights_md, idx=0)
    at external/mkl_dnn_acl_compatible/include/oneapi/dnnl/dnnl.hpp:4208
#1  0x0000ffffb7845774 in dnnl::matmul::primitive_desc::weights_desc (this=<optimized out>)
    at external/mkl_dnn_acl_compatible/include/oneapi/dnnl/dnnl.hpp:11825
#2  tensorflow::(anonymous namespace)::MklMatMulPrimitive<float, float, float>::Setup (this=this@entry=0xfff7b0002470, 
    params=...) at ./tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:957
#3  0x0000ffffb784b89c in tensorflow::(anonymous namespace)::MklMatMulPrimitive<float, float, float>::MklMatMulPrimitive (
    params=..., this=0xfff7b0002470)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/stl_vector.h:97
#4  tensorflow::(anonymous namespace)::MklMatMulPrimitiveFactory<float, float, float, float>::Get (params=..., 
    do_not_cache=do_not_cache@entry=false) at ./tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:1102
#5  0x0000ffffb784f3b8 in tensorflow::MklEinsumHelper::MKLContractOperands<Eigen::ThreadPoolDevice, float> (
    ctx=ctx@entry=0xfff7bf7fbb28, inputs=..., swap_free_and_contract=..., output=output@entry=0xfff7bf7fb460)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:421
#6  0x0000ffffb785659c in tensorflow::MklEinsum<Eigen::ThreadPoolDevice, float>::Compute (this=<optimized out>, 
    ctx=0xfff7bf7fbb28) at external/com_google_absl/absl/types/span.h:189
#7  0x0000ffffbb1eb2a8 in tensorflow::ThreadPoolDevice::Compute (this=0xaaaad3007c40, op_kernel=0xaaaad30a7d00, 
    context=0xfff7bf7fbb28) at tensorflow/core/common_runtime/threadpool_device.cc:185
#8  0x0000ffffbb168f30 in tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::ProcessSync (
    this=this@entry=0xaaaad30a3bd0, item=..., params=<optimized out>, outputs=outputs@entry=0xfff7bf7fbd08, 
    stats=stats@entry=0x0) at external/local_tsl/tsl/platform/profile_utils/cpu_utils.h:77
#9  0x0000ffffbb16f4e8 in tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::ProcessInline (
    this=this@entry=0xaaaad30a3bd0, inline_ready=inline_ready@entry=0xfff7bf7fbfe8, scheduled_nsec=scheduled_nsec@entry=0)
    at external/com_google_absl/absl/container/internal/inlined_vector.h:403
#10 0x0000ffffbb16fac0 in tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::Process (
    this=0xaaaad30a3bd0, tagged_node=..., scheduled_nsec=0) at external/com_google_absl/absl/container/inlined_vector.h:750
#11 0x0000ffffbb1528f0 in std::__invoke_impl<void, void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*&)(tensorflow::SimplePropagatorState::TaggedNode const&, long), tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>*&, tensorflow::SimplePropagatorState::TaggedNode&, long&> (__t=@0xfff7b0001db0: 0xaaaad30a3bd0, 
    __f=
    @0xfff7b0001d90: (void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*)(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState> * const, const tensorflow::SimplePropagatorState::TaggedNode &, long)) 0xffffbb16fa50 <tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::Process(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::TaggedNode const&, int64_t)>)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:89
#12 std::__invoke<void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*&)(tensorflow::SimplePropagatorState::TaggedNode const&, long), tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>*&, tensorflow::SimplePropagatorState::TaggedNode&, long&> (__fn=
    @0xfff7b0001d90: (void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*)(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState> * const, const tensorflow::SimplePropagatorState::TaggedNode &, long)) 0xffffbb16fa50 <tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::Process(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::TaggedNode const&, int64_t)>)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:95
#13 std::_Bind<void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>*, tensorflow::SimplePropagatorState::TaggedNode, long int))(const tensorflow::SimplePropagatorState::TaggedNode&, long int)>::__call<void, 0, 1, 2> (__args=..., this=0xfff7b0001d90)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/functional:416
#14 std::_Bind<void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>*, tensorflow::SimplePropagatorState::TaggedNode, long int))(const tensorflow::SimplePropagatorState::TaggedNode&, long int)>::operator()<> (this=0xfff7b0001d90)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/functional:499
#15 operator() (this=0xfff7b0001d90) at tensorflow/core/common_runtime/executor.cc:486
#16 std::__invoke_impl<void, tensorflow::(anonymous namespace)::ExecutorState<PropagatorStateType>::RunTask<std::_Bind<void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>*, tensorflow::SimplePropagatorState::TaggedNode, long int))(const tensorflow::SimplePropagator--Type <RET> for more, q to quit, c to continue without paging--
State::TaggedNode&, long int)> >::<lambda()>&> (__f=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60
#17 std::__invoke_r<void, tensorflow::(anonymous namespace)::ExecutorState<PropagatorStateType>::RunTask<std::_Bind<void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>*, tensorflow::SimplePropagatorState::TaggedNode, long int))(const tensorflow::SimplePropagatorState::TaggedNode&, long int)> >::<lambda()>&> (__fn=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:110
#18 std::_Function_handler<void(), tensorflow::(anonymous namespace)::ExecutorState<PropagatorStateType>::RunTask<std::_Bind<void (tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>::*(tensorflow::(anonymous namespace)::ExecutorState<tensorflow::SimplePropagatorState>*, tensorflow::SimplePropagatorState::TaggedNode, long int))(const tensorflow::SimplePropagatorState::TaggedNode&, long int)> >::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:291
#19 0x0000aaaaaeae4f0c in std::function<void ()>::operator()() const (this=<optimized out>)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:248
#20 0x0000ffffbbad38d0 in tsl::thread::EigenEnvironment::ExecuteTask (t=..., this=0xaaaad2fc7258)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:421
#21 Eigen::ThreadPoolTempl<tsl::thread::EigenEnvironment>::WorkerLoop (this=0xaaaad2fc7250, thread_id=<optimized out>)
    at external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/ThreadPool/NonBlockingThreadPool.h:331
#22 0x0000ffffbbad3920 in Eigen::ThreadPoolTempl<tsl::thread::EigenEnvironment>::ThreadPoolTempl(int, bool, tsl::thread::EigenEnvironment)::{lambda()#1}::operator()() const (this=<optimized out>)
    at external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/ThreadPool/NonBlockingThreadPool.h:61
#23 std::__invoke_impl<void, Eigen::ThreadPoolTempl<tsl::thread::EigenEnvironment>::ThreadPoolTempl(int, bool, tsl::thread::EigenEnvironment)::{lambda()#1}&>(std::__invoke_other, Eigen::ThreadPoolTempl<tsl::thread::EigenEnvironment>::ThreadPoolTempl(int, bool, tsl::thread::EigenEnvironment)::{lambda()#1}&) (__f=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60
#24 std::__invoke_r<void, Eigen::ThreadPoolTempl<tsl::thread::EigenEnvironment>::ThreadPoolTempl(int, bool, tsl::thread::EigenEnvironment)::{lambda()#1}&>(void&&, (Eigen::ThreadPoolTempl<tsl::thread::EigenEnvironment>::ThreadPoolTempl(int, bool, tsl::thread::EigenEnvironment)::{lambda()#1}&)...) (__fn=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:110
#25 std::_Function_handler<void (), Eigen::ThreadPoolTempl<tsl::thread::EigenEnvironment>::ThreadPoolTempl(int, bool, tsl::thread::EigenEnvironment)::{lambda()#1}>::_M_invoke(std::_Any_data const&) (__functor=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:291
#26 0x0000aaaaaeae4f0c in std::function<void ()>::operator()() const (this=<optimized out>)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:248
#27 0x0000ffffbbad3964 in tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}::operator()() const (
    __closure=0xaaaad30b5890) at external/local_tsl/tsl/platform/threadpool.cc:75
#28 std::__invoke_impl<void, tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&>(std::__invoke_other, tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&) (__f=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60
#29 std::__invoke<tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&>(std::__invoke_result&&, (tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&)...) (__fn=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:95
#30 std::invoke<tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&>(std::invoke_result&&, (tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&)...) (__fn=...)
    at /dt10/usr/lib/gcc/aarch64-unknown-linux-gnu/10/../../../../include/c++/10/functional:88
#31 absl::lts_20230802::internal_any_invocable::InvokeR<void, tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&, , void>(tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&) (f=...)
    at external/com_google_absl/absl/functional/internal/any_invocable.h:132
#32 0x0000ffffbbad3998 in absl::lts_20230802::internal_any_invocable::RemoteInvoker<false, void, tsl::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}&>(absl::lts_20230802::internal_any_invocable::TypeErasedState*) (
    state=<optimized out>) at external/com_google_absl/absl/functional/internal/any_invocable.h:359
#33 0x0000ffffbb4c45a0 in absl::lts_20230802::internal_any_invocable::Impl<void ()>::operator()() (this=0xaaaad30b5900)
    at external/com_google_absl/absl/functional/internal/any_invocable.h:868
#34 tsl::(anonymous namespace)::PThread::ThreadFn (params_arg=0xaaaad30b58e0)
    at external/local_tsl/tsl/platform/default/env.cc:93
--Type <RET> for more, q to quit, c to continue without paging--
#35 0x0000ffffae14d624 in start_thread () from /lib/aarch64-linux-gnu/libpthread.so.0
#36 0x0000ffffadd7462c in ?? () from /lib/aarch64-linux-gnu/libc.so.6
(gdb) 
```

**Comment by vpirogov** [comment]
+@milpuz01

**Comment by mdfaijul** [comment]
@milpuz01 Could you please help fixing the failure?

**Comment by milpuz01** [comment]
Hi @mdfaijul, 

as @elfringham identified the failure comes in `Setup()` method in `MklMatMulPrimtive` class when you are trying to set `b_mem` in `context_` by reading weight description from `prim_desc` that still wasn't identified. We do it this way on AArch64 so that we can set weights as `any` as we need to reorder them in order to select the most optimal kernel from Arm Compute Library (more information on this approach are available in this PR: https://github.com/tensorflow/tensorflow/pull/61171). To fix the failure you would need to set `b_mem` later when `prime_desc` is actually created (between lines 1052 and 1059 in your PR). 

One way how to do it is with [this](https://github.com/tensorflow/tensorflow/files/14227019/58204_aarch64_fix.patch) patch that @elfringham tested and those failing tests pass now on AArch64. 

Hope this helps.

Best,
Milos.

**Comment by mdfaijul** [comment]
@milpuz01 Thanks! I will take a look.
